### PR TITLE
Single resolver argument for strawberry.field

### PR DESCRIPTION
**Pull request by BryceBeagle**
## Description
Previously, `strawberry.field` had redundant arguments for the resolver, one for when `strawberry.field` was used as a decorator, and one for when it was used as a function. This PR seeks to unify them.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

#438 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)

File: strawberry/field.py
Comment (BryceBeagle) [inline strawberry/field.py:14]: Discussion: This is no longer the camelcase-d resolver name. I figure this is actually preferable, as this error will occur at startup. The camelcase name would only be helpful if this was a runtime error

File: strawberry/arguments.py
Comment (BryceBeagle) [inline strawberry/arguments.py:56]: Discussion: This is no longer the camelcase-d resolver name. I figure this is actually preferable, as this error will occur at startup. The camelcase name would only be helpful if this was a runtime error

This also means this function now only needs one argument

File: strawberry/types/type_resolver.py
Comment (BryceBeagle) [inline strawberry/types/type_resolver.py:281]: This actually doesn't get hit. An exception (a TypeError with reasonable detail) is thrown in `type_resolver._get_fields` when processing the dataclass. 

Should we just use that exception, or do we want to convert it to a Strawberry exception?

File: strawberry/types/type_resolver.py
Comment (BryceBeagle) [inline strawberry/types/type_resolver.py:313]: Why not just use `dataclasses.MISSING`?

File: strawberry/types/type_resolver.py
Comment (BryceBeagle) [reply strawberry/types/type_resolver.py:281]: (note that WindowsError is a fill-in, and won't be in the final commit)

File: strawberry/field.py
Comment (BryceBeagle) [inline strawberry/field.py:57]: This breaks when `resolver` is a classmethod. Apparently they are immutable? Not sure how to address...

File: strawberry/types/type_resolver.py
Comment (patrick91) [reply strawberry/types/type_resolver.py:313]: I didn't think of that, we can change it 😊

File: tests/schema/test_resolvers.py
Comment (patrick91) [inline tests/schema/test_resolvers.py:214]: we need to add tests to make sure that we use the type defined on the class, and that we throw when the types are different (between resolvers and the class field) :)

File: strawberry/field.py
Comment (BryceBeagle) [inline strawberry/field.py:57]: We probably want to use something along the lines of
```python
if inspect.ismethod(resolver) and resolver.__self__ is cls:
    ...
```

There are other things that could, in theory, cause an AttributeError?

Comment (BryceBeagle) [reply strawberry/field.py:57]: Also, does this work with the `hasattr(field, "_field_definition")` in `type_resolver._get_fields`?

Comment (patrick91) [reply strawberry/field.py:57]: > There are other things that could, in theory, cause an AttributeError?

maybe, not sure when it would happen though :)

> Also, does this work with the hasattr(field, "_field_definition") in type_resolver._get_fields?

yup :) 

image

Comment (BryceBeagle) [reply strawberry/field.py:57]: Actually it [looks like](https://stackoverflow.com/questions/7891277/why-does-setattr-fail-on-a-bound-method) `setattr` fails on _all_ bound methods, not just classmethods... so I think all we need is 
```python
inspect.ismethod(resolver)
```

Comment (BryceBeagle) [reply strawberry/field.py:57]: > yup :)

huh very interesting... I didn't know there was a passthrough getattr of sorts on bound methods

Comment (patrick91) [reply strawberry/field.py:57]: I wonder if we should check if a method is classmethod though, if we assign a bound that's not static or a classmethod then we might have a misusage, no?

Comment (BryceBeagle) [reply strawberry/field.py:57]: I think someone could pass a bound instance method as the resolver, right?

```python
class ResolverClass:
    _state: SomeState
    def resolve_func(self, info, arg) -> int:
        return self.do_something(self._state, info, arg)

resolver_class = ResolverClass()
```
```python
@strawberry.type
class Query:
    resolver_class.resolve_func here is a bound method
    query: int = strawberry.field(resolver=resolver_class.resolve_func)
```

Comment (patrick91) [reply strawberry/field.py:57]: oh, good point. I wonder what happens in that case with self though. This is something for later I guess (we can add an issue and a test for this), not sure it is going to be common anyway 💯

File: strawberry/types/type_resolver.py
Comment (BryceBeagle) [reply strawberry/types/type_resolver.py:325]: Will address in another PR

File: tests/schema/test_resolvers.py
Comment (BryceBeagle) [reply tests/schema/test_resolvers.py:214]: How strict do we want to be? Should any of these be valid?
```python
@strawberry.type
class Query:

    @strawberry.interface
    class Cheese:
        name: str
    @strawberry.type
    class Swiss(Cheese):
        canton: str

    def int_resolver() -> int: ...
    def swiss_resolver() -> Swiss: ...
    def uuid_resolver() -> UUID: ...
    
    my_float: float = strawberry.field(resolver=int_resolver)
    my_cheese: Cheese = strawberry.field(resolver=swiss_resolver)
    my_id: strawberry.ID = strawberry.field(resolver=uuid_resolver)
```

I think we already discussed the float/int example and determined that it should throw an exception. What about the others?

File: strawberry/types/type_resolver.py
Comment (BryceBeagle) [inline strawberry/types/type_resolver.py:325]: This check, and the one below, are currently swallowed by graphql.GraphQLSchema, which just returns a simple TypeError with our error's message.

I wonder if it would be best/possible to attempt calling `type_resolver._get_fields` _before_ building the schema?

File: tests/schema/test_resolvers.py
Comment (BryceBeagle) [reply tests/schema/test_resolvers.py:214]: Discussed on Discord. Current plan is to do the type matching verification in another PR. Will start by enforcing _exact_ matches, potentially being more lenient afterwards.

File: strawberry/types/type_resolver.py
Comment (patrick91) [reply strawberry/types/type_resolver.py:325]: oh really? that's annoying. I think maybe we can move the check when we run the decorator?

Comment (BryceBeagle) [reply strawberry/types/type_resolver.py:325]: Well the issue is that `strawberry.field` doesn't know if it's being used as a decorator, or just with `resolver` as the only argument

As far as the `strawberry.field` can tell, these are equivalent, but only the first is valid:

```python
@strawberry.type
class Query:
    def my_func(): ...
    my_field: int = strawberry.field(resolver=my_func)

    @strawberry.field
    def my_func():
        ...
```

It doesn't know if the typing will/will not be provided externally on the field

File: strawberry/types/type_resolver.py
Comment (BryceBeagle) [reply strawberry/types/type_resolver.py:357]: Created issue #439

File: strawberry/types/type_resolver.py
Comment (BryceBeagle) [reply strawberry/types/type_resolver.py]: Issue already tracked in #396 . Will address in future PR

File: strawberry/field.py
Comment (BryceBeagle) [reply strawberry/field.py]: I added a test for bounded instance method resolvers... Is there anything further that we wanted to investigate in another issue?

File: strawberry/types/type_resolver.py
Comment (patrick91) [inline strawberry/types/type_resolver.py:304]: can we add a reference to the issue for this? 😊

File: strawberry/types/type_resolver.py
Comment (patrick91) [inline strawberry/types/type_resolver.py:381]: I'll double check this with fresh eyes tomorrow, but so far really good!

**Review by patrick91 — COMMENTED** [review]
Added a first review, this is great work 👏

File: strawberry/type.py
Comment (BryceBeagle) [reply strawberry/type.py:40]: Oops and this doesn't even work if the pattern doesn't match -- I forgot to re-raise 🤦‍♂️

File: tests/types/test_resolvers.py
Comment (BryceBeagle) [reply tests/types/test_resolvers.py:90]: It's actually PyCharm... but I didn't want to leave a PyCharm-specific pragma to ignore it. Turns out I can ignore the flake8 error code and PyCharm understands it.

I do something similar [here](https://github.com/strawberry-graphql/strawberry/blob/3a4db59859ff6e4ae5900c5e820afab7f355235b/tests/schema/test_resolvers.py#L251), but that's to get PyCharm to quit complaining that the function could be static. There's no flake8 analog afaik. In hindsight, it may be too much...

Thoughts?

Comment (patrick91) [reply tests/types/test_resolvers.py:90]: yeah, I wonder why pychamr complains about self not being used, feels a bit strange, since it is a required param on classmethod :)

Comment (BryceBeagle) [reply tests/types/test_resolvers.py:90]: It's not exactly complaining that it's unused... moreso saying that the function should be static _because_ it's not used

Comment (BryceBeagle) [reply tests/types/test_resolvers.py:90]: For this test, I'm _specifically_ testing against instance methods, so it felt kinda off to have PyCharm insist that the function become static

**Comment by BryceBeagle** [comment]
Should be ready for review again

File: strawberry/types/type_resolver.py
Comment (patrick91) [inline strawberry/types/type_resolver.py:319]: when would this happen? if it doesn't I'd skip this check 😊 
we can add a `assert field_definition.base_resolver` instead, so it is clear there was a bug :)

File: tests/types/test_resolvers.py
Comment (patrick91) [reply tests/types/test_resolvers.py:37]: I'm going to check this tonight or tomorrow :)

File: tests/types/test_resolvers.py
Comment (patrick91) [inline tests/types/test_resolvers.py:91]: nit: This reads better I think :)

```suggestion
    assert e.value.args[0] == (
```

**Review by patrick91 — APPROVED** [review]
Awesome work on this! I added a couple of minor comments, but nothing major.

I just want to check if we can raise `MissingReturnAnnotationError` as early as possible.

I think we can merge this after updating the minor comments 😊

File: strawberry/type.py
Comment (patrick91) [inline strawberry/type.py:71]: @BryceBeagle what if we add some safety checks here? At least to check if all fields have return types. Would be amazing to check that before we do the schema generation, as the schema generation shouldn't have to deal with missing types :)

File: tests/types/test_resolvers.py
Comment (BryceBeagle) [reply tests/types/test_resolvers.py:90]: I dunno.... I find the single-item tuple syntax to be pretty bad...

I'll change it though to be consistent

Comment (patrick91) [reply tests/types/test_resolvers.py:90]: I just release in other places we check the whole tuple, which looks nicer 😊

Comment (BryceBeagle) [reply tests/types/test_resolvers.py:90]: Sorry, but I'm not sure what this comment is supposed to mean 😅

File: strawberry/types/type_resolver.py
Comment (BryceBeagle) [reply strawberry/types/type_resolver.py:319]: If something else in the future ever calls `_get_fields` I figured it would be helpful to catch that issue here too

I'm a little torn here... The [entire `else` clause](https://github.com/strawberry-graphql/strawberry/blob/246d23d6998a33342186e9f7caa6ef96c422ec9d/strawberry/types/type_resolver.py#L315-L324) this is in currently does nothing but identify problems with the FieldDefinition. The [other item in here](https://github.com/strawberry-graphql/strawberry/blob/246d23d6998a33342186e9f7caa6ef96c422ec9d/strawberry/types/type_resolver.py#L321-L324) is also a check, but this time it _can_ currently get hit. It feels weird to have one be an `assert` and the other raise an exception, does it not?

File: tests/types/test_resolvers.py
Comment (patrick91) [reply tests/types/test_resolvers.py:90]: this: https://github.com/strawberry-graphql/strawberry/blob/3a4b6ec0effd3c86db781059c77618651f9d71dc/tests/types/test_resolvers.py#L34-L36 :)

as opposed to sue my suggestion above with args[0] :)

File: strawberry/types/type_resolver.py
Comment (patrick91) [reply strawberry/types/type_resolver.py:319]: maybe, I just think they are different cases so it might be ok to have different ways to fail. One happens as a user error while using resolver, the other one, not sure :)

File: tests/types/test_resolvers.py
Comment (BryceBeagle) [reply tests/types/test_resolvers.py:90]: Oh I didn't even notice that the suggestion you proposed _didn't_ do that, cause you did it everywhere else and I guess I just glossed over it. That's why I mentioned the single-item tuple syntax -- that trailing comma is really easy to miss haha.

To clarify, you're suggesting I use the literal tuple (second example), instead of the indexing pattern (first example), right? I'll change it to whichever you want 👍.

File: strawberry/types/type_resolver.py
Comment (BryceBeagle) [reply strawberry/types/type_resolver.py:319]: up to you 👍

File: tests/types/test_resolvers.py
Comment (patrick91) [reply tests/types/test_resolvers.py:90]: let's do this then 😊 https://github.com/strawberry-graphql/strawberry/blob/3a4b6ec0effd3c86db781059c77618651f9d71dc/tests/types/test_resolvers.py#L34-L36

File: strawberry/types/type_resolver.py
Comment (patrick91) [reply strawberry/types/type_resolver.py:319]: let's keep it for now, I think if we do this: https://github.com/strawberry-graphql/strawberry/pull/429#discussion_r481392751

we might be good :)

**Pull request merged**
